src/sqlfluff/core/parser/lexer.py

```diff
@@ -254,6 +255,242 @@ def search(self, forward_string: str) -> Optional[Tuple[int, int]]:
         return None
 
 
+def _generate_template_loop_segments(
+    source_slice: slice,
+    last_source_slice: slice,
+    templated_idx: int,
+    templated_file: TemplatedFile,
+) -> List[RawSegment]:
+    """Detect when we've gone backward in the source.
+
+    NOTE: If it's the _same_ slice then don't insert a marker
+    because we're probably just within a single templated
+    section.
+    """
+    if (
+        last_source_slice
+        and last_source_slice.stop > source_slice.start
+        and last_source_slice.stop != source_slice.stop
+        and last_source_slice != source_slice
+    ):
+        # If we have, insert a loop marker to reflect that.
+        lexer_logger.debug("      Backward jump detected. Inserting Loop Marker")
+        # TemplateLoops should usually have a dedent before
+        # and an indent after.
+        return [
+            Dedent(
+                is_template=True,
+                pos_marker=PositionMarker.from_point(
+                    last_source_slice.stop,
+                    templated_idx,
+                    templated_file,
+                ),
```

Comment:
Nitpick: Since all 3 segments have the same `PositionMarker`, could we reuse or copy it rather than repeating these 5 lines each time?

Reply:
good point